File: src/challenges/go/problems/array_strings_are_equal_test.go
// https://leetcode.com/problems/check-if-two-string-arrays-are-equivalent

package problems

import (
	"strings"
	"testing"

	"github.com/dobermanch/leetcode/core"
)

type ArrayStringsAreEqual struct{}

func TestArrayStringsAreEqual(t *testing.T) {
	gen := core.TestSuite[ArrayStringsAreEqual]{}
	gen.Add(func(tc *core.TestCase) {
		tc.Param([]string{"ab", "c"}).Param([]string{"a", "bc"}).Result(true)
	}).Add(func(tc *core.TestCase) {
		tc.Param([]string{"a", "cb"}).Param([]string{"ab", "c"}).Result(false)
	}).Add(func(tc *core.TestCase) {
		tc.Param([]string{"abc", "d", "defg"}).Param([]string{"abcddefg"}).Result(true)
	}).Run(t)
}

func (ArrayStringsAreEqual) Solution1(word1 []string, word2 []string) bool {
    return strings.Join(word1, "") == strings.Join(word2, "")
}

func (ArrayStringsAreEqual) Solution2(word1 []string, word2 []string) bool {
    build := func(words []string) string {
        var builder strings.Builder
        for _, word := range words {
            builder.WriteString(word);
        }

        return builder.String();
    }

    return build(word1) == build(word2);
}

File: src/python/challenges/problems/array_strings_are_equal_test.py
# https://leetcode.com/problems/check-if-two-string-arrays-are-equivalent

from core.problem_base import *

class ArrayStringsAreEqual(ProblemBase):
    def Solution1(self, word1: list[str], word2: list[str]) -> bool:        
        return "".join(word1) == "".join(word2)
    
    def Solution2(self, word1: list[str], word2: list[str]) -> bool:
        def build(words):
            builder = ""
            for word in words:
                builder += word

            return builder

        return build(word1) == build(word2)

if __name__ == '__main__':
    TestGen(ArrayStringsAreEqual) \
        .Add(lambda tc: tc.Param(["ab", "c"]).Param(["a", "bc"]).Result(True)) \
        .Add(lambda tc: tc.Param(["a", "cb"]).Param(["ab", "c"]).Result(False)) \
        .Add(lambda tc: tc.Param(["abc", "d", "defg"]).Param(["abcddefg"]).Result(True)) \
        .Run()
